### ISSUE TITLE
Multiple writes saturate the result towards all-bits-set

### ISSUE THREAD

**Issue by vorner** [Fraser999/Hash-Hasher#1]
Hello

I've noticed one thing. If there are multiple calls to `write` during the lifetime of the hasher, a `1` bit stays set forever if it is set.

Why is this a problem? Because implementation of some `Hash` (even auto-generated) call write multiple times. For example, a `hash` of a slice will first hash the length, then the values. A fully-random `[u8]` slice with length 255 will always have all one byte made „ineffective“ by encoding this length.

May I suggest using XOR instead of OR in there?

**Comment by vorner** [comment]
A little demonstration what I mean:

```rust
use hash_hasher::HashHasher;
use std::fmt::Debug;
use std::hash::{Hash, Hasher};

fn x<H: Debug + Hash>(h: H) {
    let mut hasher = HashHasher::default();
    h.hash(&mut hasher);
    println!("{:?}\t{:b}", h, hasher.finish());
}

#[derive(Copy, Clone, Debug, Hash)]
struct A([u8; 8]);

#[derive(Copy, Clone, Debug, Hash)]
struct B(&'static [[u8; 8]]);

fn main() {
    x(42usize);
    x(&[1u8, 2, 3, 4, 5, 6, 7, 8] as &[u8]);
    x(&[2u8, 1, 3, 4, 5, 6, 7, 8] as &[u8]);
    x(&[2u8, 1, 3, 4, 5, 6, 7, 7] as &[u8]);
    x(A([1, 2, 3, 4, 5, 6, 7, 8]));
    x(A([0, 0, 0, 0, 0, 0, 0, 0]));
    x(A([8, 0, 0, 0, 0, 0, 0, 0]));
    x(B(&[
        [255, 0, 0, 0, 0, 0, 0, 0],
        [0, 255, 0, 0, 0, 0, 0, 0],
        [0, 0, 255, 0, 0, 0, 0, 0],
        [0, 0, 0, 255, 0, 0, 0, 0],
        [0, 0, 0, 0, 255, 0, 0, 0],
        [0, 0, 0, 0, 0, 255, 0, 0],
        [0, 0, 0, 0, 0, 0, 255, 0],
        [0, 0, 0, 0, 0, 0, 0, 255],
        [16, 0, 0, 0, 0, 0, 0, 255],
        [16, 0, 0, 23, 0, 0, 0, 255],
    ]));
    x(0u64);
}
```

For example, the all-zeroes A thing still has a bit set, that says it's 8 elements long and has the same hash as the one starting with 8.

**Comment by Fraser999** [comment]
Thanks very much for this report.

I've fixed the issue as per your recommendation in https://github.com/Fraser999/Hash-Hasher/commit/2de7c1ac30c2700cad11d669a72c3f885484fd59 and published the crate as version 2.0.0.

**Comment by vorner** [comment]
Thank you ☺
